### PR TITLE
Two ReactART TODOs implemented on Android

### PR DESCRIPTION
**Pull request by tepamid**
Implemented 2 TODOs from ReactART for Android:
- TODO(7255985): Use TextureView and pass Surface from the view to draw on it asynchronously instead of passing the bitmap (which is inefficient especially in terms of memory usage)
- TODO(6352067): Support dashes in ARTShape

We use ReactNativeART in our Android project.
1. Our app crashes sometimes on large screen smartphones with OutOfMemoryError. Crashes happen in ARTSurfaceShadowNode where TODO(7255985) was suggested in a comment in order to use memory more efficiently.
2. We needed dashes for drawing on ARTSurface.

**Test plan (required)**

I attach a screenshot of our app which shows dashed-lines and two ARTSurfaces on top of each other rendering exactly the same as in the pervious implementation of ARTSurface.
screenshot_2016-08-19-16-45-43

### CODE REVIEW

**Comment by facebook-github-bot-7** [comment]
By analyzing the blame information on this pull request, we identified @spicyj and @emilsjolander to be potential reviewers.

**Comment by facebook-github-bot-7** [comment]
Thank you for your pull request and welcome to our community. We require contributors to sign our Contributor License Agreement, and we don't seem to have you on file. In order for us to review and merge your code, please sign up at https://code.facebook.com/cla - and if you have received this in error or have any questions, please drop us a line at cla@fb.com. Thanks!

**Comment by facebook-github-bot-5** [comment]
@tepamid updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9486/files/47333292873c899ab8e94359e1b6390a5279864d..163dcaaa3cb2a40cf4d3f4c3d1b6650546d8bd0f)

**Comment by facebook-github-bot-5** [comment]
@tepamid updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9486/files/163dcaaa3cb2a40cf4d3f4c3d1b6650546d8bd0f..b368f4b4475b5ae8a3ca02db358b07fee34e00b3)

**Comment by facebook-github-bot-3** [comment]
@tepamid updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9486/files/b368f4b4475b5ae8a3ca02db358b07fee34e00b3..97dfe1b2494265616a8828a47d63bf0bd06aac88)

**Comment by facebook-github-bot-4** [comment]
@tepamid updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9486/files/97dfe1b2494265616a8828a47d63bf0bd06aac88..ff0047bcb45606df6310ceb0450a0131213275d6)

**Comment by facebook-github-bot-4** [comment]
@tepamid updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9486/files/ff0047bcb45606df6310ceb0450a0131213275d6..ac9622f802d2748eb2f62f1db8be356d3f734b43)

**Comment by facebook-github-bot-4** [comment]
@tepamid updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9486/files/9dad9800b3a121142b2bea99169bc02f33634aeb..8a1de6109b97bd1b94a36549cf8d73ad309ba4a0)

**Comment by facebook-github-bot-9** [comment]
@tepamid updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9486/files/8a1de6109b97bd1b94a36549cf8d73ad309ba4a0..c516125f7997f6006f24e722fde3ea8ec8aa1078)

**Comment by tepamid** [comment]
@sebmarkbage, when will you be able to check my PR?

**Comment by sebmarkbage** [comment]
Sorry, I'll take a look next week.

**Comment by tepamid** [comment]
@sebmarkbage, a friendly reminder.

File: ReactAndroid/src/main/java/com/facebook/react/views/art/ARTSurfaceViewShadowNode.java [hunk 1]
```diff
@@ -34,23 +42,46 @@ public boolean isVirtualAnchor() {
...
+  }
+
+  public void onSurfaceTextureAvailable(SurfaceTexture surface, int width, int height) {
+      mSurface = new Surface(surface);
   }
+
+  public boolean onSurfaceTextureDestroyed(SurfaceTexture surface) {
+      mSurface = null;
+      return true;
+  }
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/views/art/ARTSurfaceViewShadowNode.java]: Small nit: only two spaces of indentation please.

File: ReactAndroid/src/main/java/com/facebook/react/views/art/ARTSurfaceViewShadowNode.java [hunk 2]
```diff
@@ -34,23 +42,46 @@ public boolean isVirtualAnchor() {
...
-    // TODO(7255985): Use TextureView and pass Surface from the view to draw on it asynchronously
-    // instead of passing the bitmap (which is inefficient especially in terms of memory usage)
-    Bitmap bitmap = Bitmap.createBitmap(
-        (int) getLayoutWidth(),
-        (int) getLayoutHeight(),
-        Bitmap.Config.ARGB_8888);
-    Canvas canvas = new Canvas(bitmap);
+  private void drawOutput() {
+    if (mSurface == null) {
+      markChildrenUpdatesSeen(this);
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/views/art/ARTSurfaceViewShadowNode.java]: Why is this needed?

**Comment by mkonicek** [comment]
Thanks for doing this!

I haven't used these APIs in the past - @foghina Are you familiar with `TextureView` and related APIs or do you know anyone who is? :)

File: ReactAndroid/src/main/java/com/facebook/react/views/art/ARTSurfaceViewShadowNode.java (hunk 1 above)
Comment (tepamid) [inline ReactAndroid/src/main/java/com/facebook/react/views/art/ARTSurfaceViewShadowNode.java]: OK. I'll fix it.

File: ReactAndroid/src/main/java/com/facebook/react/views/art/ARTSurfaceViewShadowNode.java (hunk 2 above)
Comment (tepamid) [inline ReactAndroid/src/main/java/com/facebook/react/views/art/ARTSurfaceViewShadowNode.java]: If I do not markUpdateSeen each child in the entire Surface subtree, then ARTSurface's method onCollectExtraUpdates is called ONLY ONCE right after the app start. As I understood from the react-native sources when a child is updated it is marked "update not seen" (i.e. some ARTShape on a ARTSurface) and also it tries to mark its parent as "update not seen" only if parent doesn't have any pending updates. Otherwise if parent itself is already marked "update not seen", then an update is not propagated further to even more distant ancestors and obviously to the ultimate ancestor ARTSurface. 

That's why I need to mark all ARTSurface's children as "UpdateSeen" even if TextureView's Android.Surface is not ready for drawing.

Or is there a pre-built method in ReactNative for this?

**Comment by tepamid** [comment]
I integrated this PR into a custom build of ReactAndroid and it's used in my app's production version. It gave a significant FPS speedup while drawing on ARTSurface, so drawing performance in Android got roughly on par with its iOS counterpart.
Although sometimes IllegalArgumentException is reported while unlocking canvas. I'm investigating this issue and will update this PR accordingly in a while.

**Comment by facebook-github-bot-9** [comment]
It's been a while since the last commit was reviewed and the labels show this pull request needs review. Based on the blame information for the files in this pull request we identified @emilsjolander as a potential reviewer. Could you take a look please or cc someone with more context?

**Comment by evollu** [comment]
@tepamid it is great! but there seems to be some merge conflicts now

**Comment by robclouth** [comment]
Hey, what's the status of this?

**Comment by foghina** [comment]
@tepamid did you get rid of the random `IllegalArgumentException`s?

**Comment by tepamid** [comment]
@foghina I couldn't reproduce IllegalArgumentException myself, the code works fine. However in our production exception trackers (Raygun and Google Analytics) I rarely see that some users do face IllegalArgumentExceptions. Even more strange is that there are cases when the Surface is valid and can be successfully locked before drawing, but after drawing while unlocking the Surface IllegalArgumentException is thrown.

I googled this case and found a "holy-war" in stackoverflow whether adding try-catch around Surface locking-unlocking is a dirty hack or a reliability necessity. I personally tend to think it's a hack, but otherwise the app crashes which is even worse.

Finally, I ended with adding try-catch to handle random IllegalArgumentExceptions.

**Comment by foghina** [comment]
@tepamid nice, thanks for the summary. Do you know what happens when you catch? Does the app end up in a bad state, or does the next draw go through correctly?

**Comment by tepamid** [comment]
@foghina before I had a synthetic testcase which reproduced IllegalArgumentException in drawing on ARTSurface. An exception happened when I navigated between app screens and the screen containing ARTSurfare was unmounted. I suspect that native Android TextureView can be unmounted during drawing process putting surface handle (i.e. Canvas or Surface) into an invalid state even after the surface had been previously locked. So there won't be any more render cycles for this unmounted component and therefore lockCanvas attempts.

Also exceptions occur either in lockCanvas or unlockCanvas methods. And not during drawing shapes on a canvas.

**Comment by foghina** [comment]
Cool, thanks for the explanation! I'll go ahead and try to land this.
